XT_DAC_Audio.cpp

```diff
@@ -0,0 +1,1346 @@
+			   PlayItem->NextFillPos=0;
+
+		}
+		NextPlayItem=PlayItem->NextItem;							// move to next play item
+		if(PlayItem->Playing==false)								// If this play item completed
+		{
+			if(PlayItem->RepeatForever)
+			{
+				PlayItem->Init();						// initialise for playing again
+				PlayItem->Playing=true;
+			}
+			else
+			{
+				if(PlayItem->RepeatIdx>0)				// Repeat sound X amount of times
+				{
+					PlayItem->RepeatIdx--;
+					PlayItem->Init();						// initialise for playing again
+					PlayItem->Playing=true;
+				}
+				else
+					RemoveFromPlayList(PlayItem);		// no repeats, remove from play list
+			}
+		}
+		PlayItem=NextPlayItem;										// Set to next item
+	}
+}
+
+
+XT_PlayListItem_Class::XT_PlayListItem_Class()
+{
+	RepeatForever=false;
+	Repeat=0;
+	RepeatIdx=0;
+}
+
+// Despite being marked virtual and being overridden in desendents without these I get vtable error
+// issues which I really don't understand, sl I put these in even though never called and all is well
+uint8_t XT_PlayListItem_Class::NextByte()
+{
+	return 0;
+}
+void XT_PlayListItem_Class::Init()
+{
+}
+
+
+XT_DAC_Audio_Class::XT_DAC_Audio_Class(uint8_t TheDacPin, uint8_t TimerNo):XT_DAC_Audio_Class(TheDacPin,TimerNo,BUFFER_SIZE_DEFAULT)
+{
+	// calls other constructor as shown in function definition
+}
+
+
+XT_DAC_Audio_Class::XT_DAC_Audio_Class(uint8_t TheDacPin, uint8_t TimerNo,uint16_t PassedBufferSize)
+{
+	// Using a prescaler of 80 gives a counting frequency of 1,000,000 (1MHz) and using
+	// and calling the function every 20 counts of the frequency (the 20 below) means
+	// that we will call our onTimer function 50,000 times a second
+
+	// reserve some memory for the buffer
+	Buffer=new uint8_t[PassedBufferSize];
+	BufferSize=PassedBufferSize;					// Buffer Size is global var used in other routines that need to know size
+	for(uint32_t i=0;i<PassedBufferSize;i++)		// Set all bytes in buffer to 0x7f (DAC silence), TEB Sep-23-2019
+		Buffer[i]=0x7f;
+	FirstPlayListItem=nullptr;                      // TEB, Oct-10-2019
+	DacPin=TheDacPin;								// set dac pin to use
+	LastDacValue=0x7F;								// set to mid point
+	dacWrite(DacPin,LastDacValue);					// Set speaker to mid point, stops click at start of first sound
+	InitSineValues();								// create our table of sine values for our special circular
+													// angles of 0 - 255 "DAC Degrees". Speeds things up to if
+													// you pre-calculate
+	// Set up interrupt routine
+	timer = timerBegin(TimerNo, 80, true);          // use timer TimerNo, pre-scaler is 80 (divide by 8000), count up
+	timerAttachInterrupt(timer, &onTimer, true);    // P3= edge triggered
+	timerAlarmWrite(timer, 20, true);               // will trigger 250,000 times per second,
+	timerAlarmEnable(timer);                        // enable
+	delay(1);                             			// Allow system to settle, otherwise garbage can play for first second
```

Comment:
Given that it uses `delay()` here, we should only initialize instances of this class during setup(). Because calling `delay()` while our robot dog is operating might cause it to "freeze". If any instance of this class must be initialized during operation time, we need to modify this line.

Reply:
Can I initialize it in global? Since the instances can only be valid in setup() if they are initialized in it.
(Or I can try to use a pointer, but whether "new" can work well in esp32s remains a question.)